doc/man/gnunet-convert.1
.TH GNUNET-CONVERT "1" "20 Apr 2004" "GNUnet"
.SH NAME
gnunet-convert \- a tool to convert the GNUnet AFS database
.SH SYNOPSIS
.B gnunet\-convert
[\fIOPTIONS\fR] 
.SH DESCRIPTION
.PP

gnunet-convert can convert an existing GNUnet AFS block database from one format to another.  Current formats are "gdbm", "tdb", "bdb", "mysql" and "directory".  You must call gnunet-convert whenever you change the quota or the database type.  

.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
load config file (default: /etc/gnunet.conf)
.TP
\fB\-V\fR \fB\-\-verbose\fR
Be verbose.
.TP
\fB\-q\fR \fB\-\-quiet\fR
Be quiet.
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.SH NOTES

You cannot convert databases of GNUnet versions older than 0.5.3 with this tool.

.SH FILES
.TP
~/.gnunet/gnunet.conf
Users GNUnet configuration file
.TP
~/.gnunet/data/afs/content*
Default location of the GNUnet AFS database.
.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-search\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-gtk\fP(1), \fBgnunet\-search\fP(1), \fBgnunet\-download\fP(1), \fBgnunet\-check\fP(1), \fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/applications/afs/module/gnunet-convert.c
/*
     This file is part of GNUnet.
     (C) 2002, 2003 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/afs/module/gnunet-convert.c
 * @brief Little tool to convert content databases from format to another.
 *        Use if you change the database manager type or the bucket count.
 * @author Igor Wronsky
 * @author Christian Grothoff
 **/

#include "gnunet_util.h"
#include "gnunet_afs_esed2.h"
#include "manager.h"
#include "bloomfilter.h"
#include "manager.c"
#include "bloomfilter.c"
#include "large_file_support.c"
#include "fileindex.c"
  
static DatabaseAPI * srcHandle;
static DatabaseAPI * dstHandle;

unsigned int insertedBlocks = 0;
unsigned int failedBlocks = 0;

static int progressDot = 0;
static int be_verbose = NO;
static int be_quiet = NO;

static void addToDestination(HashCode160 * key,
			     ContentIndex *ce,
			     void * data,
			     unsigned int dataLen,
			     void * unused) {
  HashCode160 hc;
  int bucket;

  progressDot++;
  if ( ((progressDot & 255) == 0) &&
       (be_quiet == NO) ) {
    printf(".");
    fflush(stdout);    
  }

  switch (ntohs(ce->type)) {
  case LOOKUP_TYPE_CHK:
    addToBloomfilter(singleBloomFilter,
		     &ce->hash);		       
    break;
  case LOOKUP_TYPE_3HASH:
    hash(&ce->hash,
	 sizeof(HashCode160),
	 &hc);
    addToBloomfilter(singleBloomFilter,
		     &hc);
    break;
  case LOOKUP_TYPE_SUPER:
    addToBloomfilter(superBloomFilter,
		     &ce->hash);
    break;
  case LOOKUP_TYPE_SBLOCK:
    addToBloomfilter(singleBloomFilter,
	             &hc);		
    break;
  case LOOKUP_TYPE_CHKS:
    /* do nothing! */
    break;
  default:
    LOG(LOG_WARNING,
	"WARNING: encountered unexpected type %d\n",
	ntohs(ce->type));
  }
 
  bucket = computeBucket(key,
  			 dstHandle->buckets);
  if (SYSERR == dstHandle->writeContent(dstHandle->dbHandles[bucket],
					ce,
					dataLen,
					data))
    failedBlocks++;
  else
    insertedBlocks++;  
  FREENONNULL(data);
}

/**
 * Print a list of the options we offer.
 **/
static void printhelp() {
  static Help help[] = {
    HELP_CONFIG,
    HELP_HELP,
    HELP_LOGLEVEL,
    { 'q', "quiet", NULL,
      "be quiet" },
    HELP_VERSION,
    HELP_VERBOSE,
    HELP_END,
  };
  formatHelp("gnunet-convert [OPTIONS]",
	     "Convert GNUnet AFS database to different QUOTA or database type.\n"
	     "Never run gnunet-convert while gnunetd is running!",
	     help);
}

/**
 * Perform option parsing from the command line. 
 **/
static int parseCommandLine(int argc, 
			    char * argv[]) {
  int c;

  FREENONNULL(setConfigurationString("GNUNETD",
				     "_MAGIC_",
				     "YES"));
  FREENONNULL(setConfigurationString("GNUNETD",
				     "LOGFILE",
				     NULL));
  while (1) {
    int option_index = 0;
    static struct GNoption long_options[] = {
      LONG_DEFAULT_OPTIONS,
      { "verbose",  0, 0, 'V' },
      { "quiet",    0, 0, 'q' },
      { 0,0,0,0 }
    };
    
    c = GNgetopt_long(argc,
		      argv, 
		      "vhdc:nVqL:", 
		      long_options, 
		      &option_index);    
    if (c == -1) 
      break;  /* No more flags to process */
    if (YES == parseDefaultOptions(c, GNoptarg))
      continue;    
    switch(c) {
    case 'q':
      be_quiet = YES;
      break;
    case 'V':
      be_verbose = YES;
      break;
    case 'v': 
      printf("GNUnet v%s, gnunet-convert v%s\n",
	     VERSION, 
	     AFS_VERSION);
      return SYSERR;
      break;
    case 'h': 
      printhelp(); 
      return SYSERR;
      break;
    default:
      printf("Unknown option %c. Aborting.\n"
	     "Use --help to get a list of options.\n",
	     c);
      return SYSERR;
    } /* end of parsing commandline */
  }
  if (GNoptind < argc) {
    printf("Invalid arguments: ");
    while (GNoptind < argc)
      printf("%s ", argv[GNoptind++]);
    printf("\nExiting.\n");
    return SYSERR;
  }
  return OK;
}



int main(int argc, 
	 char * argv[]) {
  char * srcDb;
  char * dstDb;
  char * tmp;
  int entries;
  int newQuota;
  int * oldQuota;
  int ret;
  unsigned int i;

  if (SYSERR == initUtil(argc, argv, &parseCommandLine))
    return 0;

  newQuota = getConfigurationInt("AFS",
				 "DISKQUOTA");
  if (newQuota == 0)
    errexit("FATAL: you must specify available diskspace"
	    " in section AFS under DISKQUOTA\n"); 
  oldQuota = NULL;
  ret = stateReadContent("AFS-DISKQUOTA",
			 (void**)&oldQuota);
  if (ret != sizeof(unsigned int))
    errexit("FATAL: no conversion possible, no old database known.\n");
  tmp = NULL;
  ret = stateReadContent("AFS-DATABASETYPE",
			 (void**)&tmp);
  dstDb = getConfigurationString("AFS",
				 "DATABASETYPE");  
  if (dstDb == NULL)
    errexit("FATAL: you must specify a database type in section AFS.\n");
  if ( (ret == -1) ||
       ( (strncmp(tmp, dstDb, ret) == 0) &&
	 (newQuota == *oldQuota) ) )
    errexit("FATAL: you need to specify a different DB type "
	    "(or quota) in gnunet.conf to run gnunet-convert.\n");
  srcDb = MALLOC(ret+1);
  memcpy(srcDb, tmp, ret);
  FREENONNULL(tmp);
  srcDb[ret] = '\0';
  /* initialize old DB with old config! */
  setConfigurationInt("AFS",
		      "DISKQUOTA",
		      *oldQuota);
  FREENONNULL(oldQuota);
  FREENONNULL(setConfigurationString("AFS",
				     "DATABASETYPE",
				     srcDb));
  srcHandle = initializeDatabaseAPI(srcDb);

  /* initialize new DB with new config */
  stateWriteContent("AFS-DATABASETYPE",
		    strlen(dstDb),
		    dstDb);
  setConfigurationInt("AFS",
		      "DISKQUOTA",
		      newQuota);
  stateWriteContent("AFS-DISKQUOTA",
		    sizeof(unsigned int),
		    &newQuota);
  FREENONNULL(setConfigurationString("AFS",
				     "DATABASETYPE",
				     dstDb));

  /* FIXME: if the following call fails, we have set the new DB
     parameters in the state module but the conversion has not been
     done (and now the user is in deep shit (TM)).  We need to be able
     to prevent the state module from commiting the changes until we
     are done with everything.  The best way I can think of is to set
     a flag in state.c "don't commit", then just record all writes and
     once the flag is unset, do the writes. -- CG */

  dstHandle = initializeDatabaseAPI(dstDb);
  initBloomfilters(); /* from afs.c */
  resetBloomfilter(superBloomFilter);
  resetBloomfilter(singleBloomFilter);

  /* copy old->new */
  entries = 0;
  for (i=0;i<srcHandle->buckets;i++)
    entries += srcHandle->forEachEntryInDatabase(srcHandle->dbHandles[i],
						 &addToDestination, 
						 NULL);

  fprintf(stdout,
	  "\n==> Done processing %d entries in index "
	  "(%d converted, %d failed).\n",
	  entries,
	  insertedBlocks,
	  failedBlocks);

  /* close new, then delete old */
  doneBloomfilters();
  for (i=0;i<dstHandle->buckets;i++)
    dstHandle->doneContentDatabase(dstHandle->dbHandles[i]);
  for (i=0;i<srcHandle->buckets;i++) 
    srcHandle->deleteDatabase(srcHandle->dbHandles[i]);
  FREE(srcHandle->dbHandles);
  FREE(dstHandle->dbHandles);
  unloadDynamicLibrary(srcHandle->dynamicLibrary);
  unloadDynamicLibrary(dstHandle->dynamicLibrary);
  FREE(srcHandle);
  FREE(dstHandle);
  doneUtil();

  return 0;
}


/* end of gnunet-convert.c */
